services/keycloak/start.sh

```diff
@@ -1673,6 +1667,56 @@ function remove_billing_modifier {
+    service_api_client_id=$(/opt/jboss/keycloak/bin/kcadm.sh get -r lagoon clients?clientId=service-api --config $CONFIG_PATH)
+    if [ "$service_api_client_id" != "[ ]" ]; then
+        echo "Client service-api is already created, skipping basic setup"
+        return 0
+    fi
+    echo Creating client service-api
+    echo '{"clientId": "service-api", "publicClient": false, "standardFlowEnabled": false, "serviceAccountsEnabled": true, "secret": "'$KEYCLOAK_SERVICE_API_CLIENT_SECRET'"}' | /opt/jboss/keycloak/bin/kcadm.sh create clients --config $CONFIG_PATH -r ${KEYCLOAK_REALM:-master} -f -
+    SERVICE_API_CLIENT_ID=$(/opt/jboss/keycloak/bin/kcadm.sh get  -r lagoon clients?clientId=service-api --config $CONFIG_PATH | python -c 'import sys, json; print json.load(sys.stdin)[0]["id"]')
+    REALM_MANAGEMENT_CLIENT_ID=$(/opt/jboss/keycloak/bin/kcadm.sh get  -r lagoon clients?clientId=realm-management --config $CONFIG_PATH | python -c 'import sys, json; print json.load(sys.stdin)[0]["id"]')
+    echo Enable fine grained permissions
+    # 2 Enable fine grained admin permissions for client
+    /opt/jboss/keycloak/bin/kcadm.sh update clients/$SERVICE_API_CLIENT_ID/management/permissions --config $CONFIG_PATH -r ${KEYCLOAK_REALM:-master} -s enabled=true
+
+    CLIENT_ID=$(/opt/jboss/keycloak/bin/kcadm.sh get  -r lagoon clients?clientId=service-api --config $CONFIG_PATH | python -c 'import sys, json; print json.load(sys.stdin)[0]["id"]')
+    echo Creating groups mapper for service-api
+    echo '{"protocol":"openid-connect","config":{"full.path":"true","id.token.claim":"false","access.token.claim":"true","userinfo.token.claim":"false","claim.name":"group_membership"},"name":"Group Membership","protocolMapper":"oidc-group-membership-mapper"}' | /opt/jboss/keycloak/bin/kcadm.sh create -r ${KEYCLOAK_REALM:-master} clients/$CLIENT_ID/protocol-mappers/models --config $CONFIG_PATH -f -
+    echo Creating realm roles mapper for service-api
+    echo '{"protocol":"openid-connect","config":{"multivalued":"true","id.token.claim":"false","access.token.claim":"true","userinfo.token.claim":"false","claim.name":"realm_roles"},"name":"User Realm Roles","protocolMapper":"oidc-usermodel-realm-role-mapper"}' | /opt/jboss/keycloak/bin/kcadm.sh create -r ${KEYCLOAK_REALM:-master} clients/$CLIENT_ID/protocol-mappers/models --config $CONFIG_PATH -f -
+    echo Creating group lagoon project IDs mapper for service-api
+    echo '{"protocol":"openid-connect","config":{"id.token.claim":"false","access.token.claim":"true","userinfo.token.claim":"false","multivalued":"true","aggregate.attrs":"true","user.attribute":"group-lagoon-project-ids","claim.name":"group_lagoon_project_ids","jsonType.label":"String"},"name":"Group Lagoon Project IDs","protocolMapper":"oidc-usermodel-attribute-mapper"}' | /opt/jboss/keycloak/bin/kcadm.sh create -r ${KEYCLOAK_REALM:-master} clients/$CLIENT_ID/protocol-mappers/models --config $CONFIG_PATH -f -
+}
+
+function configure_token_exchange {
```

Comment:
As mentioned above I did initially try to do it this way, but the problem I ran into is that whichever function ran last set the token-exchange policy on that client only, and removed the token exchange policy from the other client :disappointed: 

Can you see a better way to do the same thing with less intrusive changes to the api client config?

Reply:
I haven't tested but would it also work if you just keep the `configure_token_exchange` logic but don't modify `configure_api_client`? That's basically saying 1) set token exchange for `auth-server` only 2) update token exchange for both `auth-server` and `service-api`. The `api` will know it's ready after `1` and when the policy is updated later it doesn't break anything? You can still grab the IDs of both clients, without setting the policy for both only the one time.

And if you move the logic of `configure_token_exchange` into `configure_service_api_client` you don't have to "check if policies already exist" (line 1696), because you'll know that `configure_api_client` was just run or was run previously, and you can assume if the `service-api` doesn't exist (you're already checking that in `configure_service_api_client`) you need to update the token exchange for both services.

Reply:
I refactored as you suggested, and CI is passing so :crossed_fingers: it didn't break anything.

Could you take another look when you get a chance?